### DISCUSSION TITLE
Load feathers client in service worker

### DISCUSSION THREAD

**Discussion by ericuldall** [feathersjs/feathers#2639] [Q&A]
Hello we currently use feathers client in a chrome manifest v2 extension which uses background pages. This works great because they work like the dom, every file included is available in shared scope. Unfortunately v2 is sunsetting this year and we're forced to migrate to manifest v3 which uses service workers. These do not have access to window or any other shared dom type scopes.

So as I plan to migrate our extension I'm stuck with the question of how does the feathers client fit in. Fortunately the service worker does have a mode where it can run as an ES Module. So to me the easiest solution is to see if feathers can now or will soon be available with an ESM build. Is this currently possible?

Or if there's any other solutions to get this working in a service worker i'm all ears. Just getting my head around the change and looking for some guidance.

https://developer.chrome.com/docs/extensions/mv3/service_workers/
https://web.dev/es-modules-in-sw/

tl;dr
Can feathers client be used as an ESM?

**Comment by daffl** [comment]
The Feathers client does not need a DOM and can be loaded already into a service worker (although not as an ES module - the related issue for that is https://github.com/feathersjs/feathers/issues/2423 and there is a branch for v5 at https://github.com/feathersjs/feathers/tree/esm-build but it has build issues for some reason).

**Reply by ericuldall** [reply]
I saw this issue and it seemed that it's not stable as of now. Is this production ready and does using the v5 client require any changes to the server?

**Reply by ericuldall** [reply]
Also, re: loading in a service worker not as an es module.. is there documentation on this?

**Reply by daffl** [reply]
Have you tried loading the v4 client bundle into the service worker?

**Reply by ericuldall** [reply]
I haven't. Is the bundle what we would find in the cdn? or is there a bundle in source code/npm ? Also looking at a solution using rollup to combine multiple files including dependencies into a single service worker file to allow me to have a modular workflow. Unfortunately chrome only allows one service-worker file as of now.

**Reply by daffl** [reply]
Any of the options from https://docs.feathersjs.com/api/client.html#module-loaders or https://docs.feathersjs.com/api/client.html#feathersjs-client should work fine in a service worker.
